datafusion/core/src/config.rs

```diff
@@ -17,606 +17,490 @@
-                let mut env_key = config_def.key.replace('.', "_");
-                env_key.make_ascii_uppercase();
-                match env::var(&env_key) {
-                    Ok(value) => match ScalarValue::try_from_string(
-                        value.clone(),
-                        &config_def.data_type,
-                    ) {
-                        Ok(parsed) => parsed,
-                        Err(_) => {
-                            warn!("Warning: could not parse environment variable {}={} to type {}.", env_key, value, config_def.data_type);
-                            config_def.default_value.clone()
-                        }
-                    },
-                    Err(_) => config_def.default_value.clone(),
-                }
-            };
-            options.insert(config_def.key.clone(), config_value);
+    pub fn from_env() -> Result<Self> {
+        struct Visitor(Vec<String>);
+
+        impl Visit for Visitor {
+            fn some<V: Display>(&mut self, key: &str, _: V, _: &'static str) {
+                self.0.push(key.to_string())
+            }
+
+            fn none(&mut self, key: &str, _: &'static str) {
+                self.0.push(key.to_string())
+            }
         }
-        Self { options }
-    }
 
-    /// set a configuration option
-    pub fn set(&mut self, key: &str, value: ScalarValue) {
-        self.options.insert(key.to_string(), value);
-    }
+        // Extract the names of all fields and then look up the corresponding
+        // environment variables. This isn't hugely efficient but avoids
+        // ambiguity between `a.b` and `a_b` which would both correspond
+        // to an environment variable of `A_B`
 
-    /// set a boolean configuration option
-    pub fn set_bool(&mut self, key: &str, value: bool) {
-        self.set(key, ScalarValue::Boolean(Some(value)))
-    }
+        let mut keys = Visitor(vec![]);
+        let mut ret = Self::default();
+        ret.built_in.visit(&mut keys, "datafusion", "");
 
-    /// set a `u64` configuration option
-    pub fn set_u64(&mut self, key: &str, value: u64) {
-        self.set(key, ScalarValue::UInt64(Some(value)))
-    }
+        for key in keys.0 {
+            let env = key.to_uppercase().replace('.', "_");
+            if let Some(var) = std::env::var_os(env) {
+                ret.set(&key, var.to_string_lossy().as_ref())?;
+            }
+        }
 
-    /// set a `usize` configuration option
-    pub fn set_usize(&mut self, key: &str, value: usize) {
-        let value: u64 = value.try_into().expect("convert u64 to usize");
-        self.set(key, ScalarValue::UInt64(Some(value)))
+        Ok(ret)
     }
 
-    /// set a `String` configuration option
-    pub fn set_string(&mut self, key: &str, value: impl Into<String>) {
-        self.set(key, ScalarValue::Utf8(Some(value.into())))
-    }
+    /// Returns the [`ConfigEntry`] stored within this [`ConfigOptions`]
+    pub fn entries(&self) -> Vec<ConfigEntry> {
```

Comment:
This is the mechanism by which `ConfigOptions` can be both serialized and documentation generated